flask-version/metagenomongo/module/validation.py

```diff
@@ -0,0 +1,125 @@
+    corrected_count = 0
+    corrected_items = []
+    invalid_date_messages = []
+    invalid_combobox_messages = []
+    data = df_temp.values.tolist()
+    int_dynamic_type = create_data_type_list("int", fields, options)
+    float_dynamic_type = create_data_type_list("float", fields, options)
+    # Apply corrections to data
+    for row_index, row in enumerate(data):
+        for col_index, cell in enumerate(row):
+            if isinstance(cell, str):  # Process only strings
+                original_cell = cell
+                cell = cell.strip(',; ')
+                cell = re.sub(r'\s+', ' ', cell)
+                cell = re.sub(r',\s*(\S)', r'; \1', cell)
+                
+                # Apply specific column corrections
+                if fields[col_index] == "locality":
+                    cell = re.sub(r':(?!\s)', ': ', cell)
+                elif fields[col_index] == "source_type":
+                    cell = cell.capitalize()
+                elif fields[col_index] == "if_repeated":
+                    cell = "y" if cell.lower() == "yes" else "n" if cell.lower() == "no" else cell
+                
+                if cell != original_cell:
+                    data[row_index][col_index] = cell
+                    corrected_count += 1
+                    corrected_items.append(f'Row {row_index + 1}, Column {col_index + 1}: {original_cell} -> {cell}')
+            
+            elif isinstance(cell, datetime.datetime):  # Handle datetime objects separately
+                original_cell = cell
+                cell = cell.strftime('%Y-%m-%d %H:%M:%S')
+                if cell != original_cell:
+                    data[row_index][col_index] = cell
+                    corrected_count += 1
+                    corrected_items.append(f'Row {row_index + 1}, Column {col_index + 1}: {original_cell} -> {cell}')
+    date_pattern = re.compile(r'^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2}\.\d{3}Z)?$') 
+    # Validate data
+    for row_index, row in enumerate(data):
+        for col_index, cell in enumerate(row):
+            header = fields[col_index]
+            if isinstance(cell, str):
+                if header in ["collection_date", "run_date"]:
+                    
+                    if not date_pattern.match(cell):
+                        invalid_date_messages.append(f"Invalid date in row {row_index + 1}, column '{header}': '{cell}'")
+                if header in options and options[header]['combobox_type'] == 'fix' and options[header]['options']:
+                    if cell not in options[header]['options']:
+                        invalid_combobox_messages.append(f"Invalid fixed option in row {row_index + 1}, column '{header}': '{cell}'")
+                if header in int_dynamic_type:
+                    if cell != "":
+                        if not cell.isdigit():
+                            invalid_combobox_messages.append(f"Invalid data type in row {row_index + 1}, column '{header}': '{cell}': Int")
```

Comment:
The message should say "Expected data type: int" rather than just "Int".
Same for float in the next condition.